### PR TITLE
 Signup: Add Siteless support to Plans and Domains steps

### PR DESCRIPTION
**Pull request by dylanrevisited**
#### Changes proposed in this Pull Request

* This is an updated version of #39704 , which was created a couple months ago by @mmtr 
* When the feature flag is turned on (it needs to be turned on manually right now), the site creation step will be skipped in the signup flow unless the users opt for a free plan.
* Instead of adding the flag checks directly in the fulfilled/apiRequest function for each step, two new steps were added called plans-siteless, domains-siteless, this is because we need to control in which flows we actually want to add siteless support without having a huge effect into the other steps for that flow.
* Part of #39424 and pbOQVh-cF-p2

#### Testing instructions

* Apply D43695-code and sandbox the API
* Go to `start/domains?flags=+signup/siteless-checkout/onboarding`
* Complete all the steps and select a paid plan.
* Make sure you're redirected to the checkout page and the site has not been created.
  * Please note that the checkout page won't work, this is because the needed changes are not covered in this PR and the backend changes are not ready yet.
* Make sure the signup flow remains unmodified (sites created before checkout) when turning off the feature flag by going to `/start/domains?flags=-signup/siteless-checkout/onboarding`

##### Here are the scenarios that I have validated for the **onboarding** flow.

**When signup/siteless-checkout/onboarding is on**
`/start/domains?flags=+signup/siteless-checkout/onboarding`
No Domain + Paid Plan = Go to siteless checkout
Paid Domain + Paid Plan = Go to siteless checkout
No Domain + Free Plan = Site Created + go to site home

**When signup/siteless-checkout/onboarding is off**
`/start/domains?flags=-signup/siteless-checkout/onboarding`
No Domain + Paid Plan = Site Created + Go to checkout
Paid Domain + Paid Plan = Site Created + Go to checkout
No Domain + Free Plan = Site Created + go to site home

#### Known Issues
* If you repeat the flow several times you will end up with multiple domains in the cart every time you land to the checkout page, I'll address that in a future PR.

### CODE REVIEW

**Comment by DavidRothstein** [comment]
I did an initial cursory review -- overall it made sense, but I'm a little worried that it makes the signup code (which already seems pretty complex and hard to follow, at least to me) even more complicated.

Some of that is an issue of naming (I found it confusing that with this pull request, the `createSiteWithCart` function doesn't always create a site, whereas the `addPlanToCart` function now sometimes does create a site, even though the name doesn't indicate that).  But I think I was also thrown off by the addition of the extra parallel steps: `plans-siteless` in addition to `plans`, etc.

Just looking at how the basic signup flow works in Calypso currently, it seems to roughly be this:

```
domains: {
	apiRequestFunction: createSiteWithCart,
	delayApiRequestUntilComplete: true,
},

plans: {
	apiRequestFunction: addPlanToCart,
}
```

where the `delayApiRequestUntilComplete` option (as I understand it) means that it is always delayed until after the user finishes both steps -- at that point, it will run `createSiteWithCart` followed by `addPlanToCart`.

I'm wondering if it would be possible to implement this in a simpler way by keeping most of that the same but just replacing `createSiteWithCart` with something named more like `createSiteOrCart`, which would either (a) do exactly what it does now (i.e. create a site) if the user isn't buying anything, or (b) create a siteless cart, if the user is buying something.  And then `addPlanToCart` would just have to be smart enough to add the plan to the siteless cart if there is no site yet.

Is this something you've looked into, and do you know of any reason why it wouldn't work?

**Comment by dylanrevisited** [comment]
> But I think I was also thrown off by the addition of the extra parallel steps: `plans-siteless` in addition to `plans`, etc.

I had to add the extra `-siteless` steps and the `isSiteless` argument to the existing functions instead of checking for the feature flag directly inside of them because I didn't want to modify all the flows, only a selection of them (onboarding, onboarding-plan-first, business, premium, personal). 

Moving the checks as you suggest will definitely clean up some of the diff, at least at first, but I'm not sure if some extra conditions would need to be added in order to support the rest of the flows, I'll give it a try but it'll definitely take some time since I need to validate all of them. (I'm also not sure if all of them are still in use).

Another approach that may simplify the code and give us a bit more of control is adding multiple feature flags, example: `signup/siteless-checkout/onboarding`, `signup/siteless-checkout/onboarding-plan-first`, `signup/siteless-checkout/personal` that way we can decide in which specific flows we are going to enable siteless checkout without the risk of affecting others.

**Comment by DavidRothstein** [comment]
Yeah, separate feature flags could be good, or perhaps just checking the flow name directly via a hardcoded list.  I guess I was thinking roughly something like this:

```
export function createSiteOrCart( callback, dependencies, stepData, reduxStore ) {
	if ( ! isEnabled( 'signup/siteless-checkout' ) || stepData.flowName !== 'onboarding' ) {
		// Fall back on the old way if not one of the enabled flows.
		return createSiteWithCart( callback, dependencies, stepData, reduxStore );
	}

	// Do the new siteless checkout stuff here.
}
```

Eventually I guess we'll want to turn siteless checkout on everywhere or almost everywhere, but it definitely makes sense to roll that out slowly.

Hopefully, `addPlanToCart()` and any others could be made smart enough to deal with either a siteless checkout or a non-siteless checkout without checking any feature flags.  I could imagine passing the cart key (either the site slug or 'no-site') along as one of the step dependencies, for example, or maybe checking the site slug directly, but I'm not sure of the details.

**Comment by dylanrevisited** [comment]
> Yeah, separate feature flags could be good, or perhaps just checking the flow name directly via a hardcoded list.

I prefer having the control using separate feature flags, especially since it will allow us to test in any env by just enabling the flag using query parameters. I have done those changes and the diff is definitely cleaner, although I decided to not rename any existing functions for now.


> Eventually I guess we'll want to turn siteless checkout on everywhere or almost everywhere, but it definitely makes sense to roll that out slowly.

Once we get there we can move to the suggested hardcoded list approach, it's just matter of changing what the `isSitelessCheckoutEnabledForFlow` function does.

> Hopefully, `addPlanToCart()` and any others could be made smart enough to deal with either a siteless checkout or a non-siteless checkout without checking any feature flags. I could imagine passing the cart key (either the site slug or 'no-site') along as one of the step dependencies, for example, or maybe checking the site slug directly, but I'm not sure of the details.

Right now it takes some conditions into account to decide what to do, but the `isSiteless` flag is more of a precautionary measure for now, I really don't want to leak any of this to other flows unless I have experimented with more of them.

File: client/signup/config/steps-pure.js [hunk 1]
```diff
@@ -166,7 +166,9 @@ export function generateSteps( {
 			stepName: 'plans',
 			apiRequestFunction: addPlanToCart,
 			dependencies: [ 'siteSlug' ],
-			providesDependencies: [ 'cartItem' ],
+			optionalIncomingDependencies: [ 'siteId', 'domainItem', 'newSiteParams' ],
```
Comment (DavidRothstein) [inline client/signup/config/steps-pure.js:169]: This is a little confusing, because if it's optional, then by definition the step does not really have it as a "dependency" :smile: 

I see why you did this, though -- you needed to pass this stuff along to the API request function, and the dependencies are really the only way to do that without a larger refactoring.

Two thoughts:
1. I would say at a minimum, if we're going to add this, let's document it in `client/signup/README.md` along with the other properties.
2. Also, an alternate suggestion, could something like this work instead?:

```
dependencies: [ 'siteSlug', 'siteId', 'domainItem', 'newSiteParams' ],
dependencyDefaults: {
  siteId: null,
  domainItem: {},
  newSiteParams: {},
},
```

The idea is that everything would still go in `dependencies` (and the step could always assume that all dependencies are present), but the code that builds the list of dependencies would use `dependencyDefaults` to fill in any missing properties before passing them along.

To me this would be easier to understand and probably more flexible for the future too.

Comment (dylanrevisited) [reply client/signup/config/steps-pure.js:169]: I'm not sold on the dependencyDefaults idea because the flow controller relies on the mere existing of the keys in the dependencies object to consider them "fulfilled", even if their value is null or undefined, I think it may also lead to some confusion because we would be basically values that do not exist in the signup progress store.

I'm more inclined to maintain `optionalIncomingDependencies` (name suggestions are welcome) or just pass the whole object of dependencies to the processing function, I don't see how the extra properties would even affect them. I'd like to know @niranjan-uma-shankar thoughts about this, since he's more familiar with the signup flows.

File: client/signup/config/flows.js [hunk 2]
```diff
@@ -56,7 +56,10 @@ function getRedirectDestination( dependencies ) {
 }
 
 function getSignupDestination( dependencies ) {
-	return `/home/${ dependencies.siteSlug }`;
+	if ( dependencies?.siteSlug !== 'no-site' ) {
+		return `/home/${ dependencies.siteSlug }`;
+	}
+	return `/checkout/thank-you/no-site`;
```
Comment (niranjan-uma-shankar) [inline client/signup/config/flows.js:62]: I am not sure if this change is necessary. Signup destination is the page where we take the user finally after payment is complete, or site has been created in the case of a free plan. In the case of siteless checkout, we still want to take the user to /home after payment is successful and not a thank you page. Could you please explain more about this change?

File: client/signup/config/steps-pure.js (hunk 1 above)
Comment (niranjan-uma-shankar) [reply client/signup/config/steps-pure.js:169]: Just an fyi that we already use `defaultDependencies`in a few places. 

https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/steps-pure.js#L202

File: client/signup/config/flows.js (hunk 2 above)
Comment (dylanrevisited) [reply client/signup/config/flows.js:62]: This is actually one of the last things that I needed to figure out, the URLs generated here are using the `dependencies` object, however we'll have a proper `siteSlug` until the checkout is completed, so another mechanism will be probably needed to redirect the user to their site home.

**Comment by niranjan-uma-shankar** [comment]
Another point is that when a user selects a paid plan and reaches the /checkout page, if they click on the delete(trash) icon in checkout, then we would need to create a site. I understand that you would be getting to this in a different PR that deals with the checkout side of the code, but just thought of noting it here.

File: client/lib/signup/step-actions/index.js [hunk 3]
```diff
@@ -283,11 +310,33 @@ export function setThemeOnSite( callback, { siteSlug, themeSlugWithRepo } ) {
...
-	const { siteSlug } = dependencies;
-	const { cartItem } = stepProvidedItems;
+	const { siteSlug, siteId, domainItem, newSiteParams } = dependencies;
+	const { cartItem, flowName, lastKnownFlow } = stepProvidedItems;
+	const flowToCheck = flowName || lastKnownFlow;
+	const isSiteless = isSitelessCheckoutEnabledForFlow( flowToCheck );
 	if ( isEmpty( cartItem ) ) {
 		// the user selected the free plan
-		defer( callback );
+		if ( isSiteless && siteId === null && isEmpty( domainItem ) && ! isEmpty( newSiteParams ) ) {
```
[the remark this answers is not in the document]
Comment (dylanrevisited) [reply client/lib/signup/step-actions/index.js]: I moved the site creation to a new step called `create-site`, that should make things easier to decouple. 

However in the process I ran into a couple issues:
1) There are several things involved to retrieve the `newSiteParams`, so i moved it to its own function to avoid duplication.
2) The cart updates during the signup are prone to race conditions, if there are several steps being processed at the same time, and both of them try to update the cart, there're high chances one of those updates will be lost, because the current process GETs the cart from the server and then POSTS the new one with the updated products. Right now it's not an issue, but it will probably be in the future, maybe it would make sense to delay the cart creation/updates.

**Comment by DavidRothstein** [comment]
>Another point is that when a user selects a paid plan and reaches the /checkout page, if they click on the delete(trash) icon in checkout to empty cart, then we would need to create a site at that point.

Ideally we would not create a site, but rather we would direct them back to the plans/domains selection step in signup instead, so they can start again and make a different choice.  (We don't know that they necessarily want a free site at that point -- just that they don't want whatever was in their cart.)

Previous internal comments about this: pbAPfg-7C-p2#comment-659

**Comment by niranjan-uma-shankar** [comment]
Thanks for the changes!

How will `newSiteParams` be stored if a paid plan is selected? I believe we will need this value to be saved so that checkout can send that value to the transactions endpoint which is responsible for creating the site.

**Comment by dylanrevisited** [comment]
> Thanks for the changes!
> 
> How will `newSiteParams` be stored if a paid plan is selected? I believe we will need save this value somewhere so that checkout can send that value to the transactions endpoint which is responsible for creating the site.

Thanks for your feedback! It has been really helpful.

We plan to store those changes in the cart to then be processed during checkout, but that will be part of another PR.

**Comment by dylanrevisited** [comment]
This was already implemented in a different way. See #45040
